src/classes/frSubscription.cls

```diff
@@ -0,0 +1,119 @@
+    public frSubscription(Map<String, Object> request) {
+        if (Boolean.valueOf(request.get('deleted'))) {
+            deleteSubscription(String.valueOf(request.get('id')));
+        } else {
+            create(request);
+        }
+    }
+    
+    public void deleteSubscription(String funraiseId) {
+        try {
+            delete [SELECT id FROM Subscription__c WHERE fr_ID__c = :funraiseId];
+        } catch (DMLException e) {
+            insert new Error__c(Error__c =
+                                'Failed to delete funraise registration for registration id ' + funraiseId +
+                                '. Exception: ' + e.getMessage());
+        }
+    }
+    
+    public void create(Map<String, Object> request) {
+        String funraiseId = String.valueOf(request.get('id'));
+        String name = String.valueOf(request.get('name'));
+        String funraiseSupporterId = String.valueOf(request.get('supporterId'));
+        List<Contact> contacts = [SELECT Id from Contact WHERE fr_Id__c = :funraiseSupporterId];
+        if(contacts.isEmpty()) {
+            insert new Error__c(Error__c =
+                                'Failed to find related record: Supporter with id ' + funraiseSupporterId+ ' for subscription id ' + funraiseId);
+            return;
+        }
+        
+        Subscription__c subscription = new Subscription__c(
+        	fr_ID__c = funraiseId,
+            Name = frUtil.truncateToFieldLength(
+                Subscription__c.Name.getDescribe(), 
+                String.valueOf(request.get('name'))
+            ), 
+            Supporter__c = contacts.get(0).Id
+        );
+        
+        String goalId = String.valueOf(request.get('campaignGoalId'));
+        if (String.isNotBlank(goalId)) {
+            List<Campaign> campaigns = [SELECT Id FROM Campaign WHERE fr_ID__c = :goalId];
+            if (campaigns.size() > 0) {
+                subscription.Campaign_Goal__c = campaigns.get(0).Id;
+            } else {
+                insert new Error__c(Error__c =
+                'Failed to find related record: campaign with id ' + goalId+ ' for subscription id ' + funraiseId);
+            }
+        }
+        
+		//transform other request fields into SF custom fields, these are not applicable
+        request.remove('id');
+        request.remove('name');
+        request.remove('supporterId');
+        request.remove('campaignGoalId');
+        request.remove('deleted');
+        
+        String regex = '([a-z])([A-Z]+)';
+        String replacement = '$1_$2';
+        for(String funraiseFieldName : request.keySet()) {
+            String salesforceFieldName = funraiseFieldName.replaceAll(regex, replacement);
+            salesforceFieldName =+ 'funraise__' + salesforceFieldName.capitalize() + '__c';
+            Schema.SObjectField field = fields.get(salesforceFieldName);
+            frUtil.write(subscription, field, funraiseFieldName, request.get(funraiseFieldName), funraiseId);
```

Comment:
Any special handling/conversion that needs to happen with the difference in types between `lastPaymentDate` and `nextPaymentDate`?

Reply:
Not needed since we've removed lastPaymentDate as a field now